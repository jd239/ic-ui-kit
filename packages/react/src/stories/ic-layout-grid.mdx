import {
  Meta,
  Stories,
  Markdown
} from "@storybook/blocks";
import layoutGridReadme from "../../../web-components/src/components/ic-layout-grid/readme.md";
import layoutGridItemReadme from "../../../web-components/src/components/ic-layout-grid-item/readme.md";
import * as IcLayoutGridStories from './ic-layout-grid.stories';

<Meta of={IcLayoutGridStories} />

<Markdown>{layoutGridReadme + layoutGridItemReadme}</Markdown>

<Stories />
