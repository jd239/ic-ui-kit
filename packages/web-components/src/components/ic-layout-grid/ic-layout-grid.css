@import "../../global/normalize.css";

:host {
  display: grid;
  width: fit-content;
  grid-template-columns: repeat(
    var(--ic-layout-grid-columns),
    var(--ic-layout-grid-col-width, 1fr)
  );
  grid-template-rows: repeat(var(--ic-layout-grid-rows), 1fr);
  gap: var(--ic-layout-grid-spacing);
  padding-top: var(--ic-layout-grid-margin);
  padding-bottom: var(--ic-layout-grid-margin);
}

:host ::slotted(*) {
  grid-column: span var(--ic-layout-grid-col-span, 1);
  grid-row: span var(--ic-layout-grid-row-span, 1);
}

:host(.ic-layout-grid-aligned-center) {
  margin-left: auto;
  margin-right: auto;
  padding-left: var(--ic-layout-grid-margin);
  padding-right: var(--ic-layout-grid-margin);
}

:host(.ic-layout-grid-aligned-left) {
  margin-left: var(--ic-layout-grid-margin);
  margin-right: auto;
  padding-right: var(--ic-layout-grid-margin);
}

:host(.ic-layout-grid-aligned-full-width) {
  margin-left: var(--ic-layout-grid-margin);
  margin-right: var(--ic-layout-grid-margin);
  width: calc(100% - (var(--ic-layout-grid-margin) * 2));
}

:host(.ic-layout-grid-no-vertical-padding) {
  padding-top: 0;
  padding-bottom: 0;
}

@media only screen and (max-width: 576px) {
  :host {
    --ic-layout-grid-margin: var(--ic-space-xs);
    --ic-layout-grid-columns: var(--ic-layout-grid-columns-xs);
  }

  :host ::slotted(*) {
    --ic-layout-grid-col-span: var(--ic-layout-grid-col-span-xs);
  }
}

@media only screen and (min-width: 577px) {
  :host {
    --ic-layout-grid-margin: var(--ic-space-md);
    --ic-layout-grid-columns: var(--ic-layout-grid-columns-sm);
  }

  :host ::slotted(*) {
    --ic-layout-grid-col-span: var(--ic-layout-grid-col-span-sm);
  }
}

@media only screen and (min-width: 769px) {
  :host {
    --ic-layout-grid-margin: var(--ic-space-md);
    --ic-layout-grid-columns: var(--ic-layout-grid-columns-md);
  }

  :host ::slotted(*) {
    --ic-layout-grid-col-span: var(--ic-layout-grid-col-span-md);
  }
}

@media only screen and (min-width: 993px) {
  :host {
    --ic-layout-grid-margin: var(--ic-space-md);
    --ic-layout-grid-columns: var(--ic-layout-grid-columns-lg);
  }

  :host ::slotted(*) {
    --ic-layout-grid-col-span: var(--ic-layout-grid-col-span-lg);
  }
}

@media only screen and (min-width: 1201px) {
  :host {
    --ic-layout-grid-margin: var(--ic-space-lg);
    --ic-layout-grid-columns: var(--ic-layout-grid-columns-xl);
  }

  :host ::slotted(*) {
    --ic-layout-grid-col-span: var(--ic-layout-grid-col-span-xl);
  }
}
