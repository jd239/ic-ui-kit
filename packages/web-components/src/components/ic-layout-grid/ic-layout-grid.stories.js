import { html } from "lit-html";

export default {
  title: "Web Components/Layout grid",
  component: "ic-layout-grid",
};

export const Default = {
  render: () => html`
    <ic-layout-grid>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Default",

  parameters: {
    layout: "fullscreen",
  },
};

export const Fluid = {
  render: () => html`
    <ic-layout-grid type="fluid">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Fluid",

  parameters: {
    layout: "fullscreen",
  },
};

export const Columns = {
  render: () => html`
    <ic-layout-grid columns="4">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Columns",

  parameters: {
    layout: "fullscreen",
  },
};

export const Rows = {
  render: () => html`
    <ic-layout-grid columns="3" rows="2">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Rows",

  parameters: {
    layout: "fullscreen",
  },
};

export const AlignedCenter = {
  render: () => html`
    <ic-layout-grid columns="4" aligned="center">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Aligned center",

  parameters: {
    layout: "fullscreen",
  },
};

export const FullWidth = {
  render: () => html`
    <ic-layout-grid columns="4" aligned="full-width">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Aligned full width",

  parameters: {
    layout: "fullscreen",
  },
};

export const FullHeight = {
  render: () => html`
    <ic-layout-grid full-height="true" columns="4">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Full height",

  parameters: {
    layout: "fullscreen",
  },
};

export const ColumnObject = {
  render: () => html`
    <ic-layout-grid>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
    <script>
      const grid = document.querySelector("ic-layout-grid");
      grid.columns = { xs: 1, sm: 2, md: 4, lg: 8, xl: 12 };
    </script>
  `,

  name: "Column object",

  parameters: {
    layout: "fullscreen",
  },
};

export const DefaultColWidth = {
  render: () => html`
    <ic-layout-grid columns="4" default-col-width="150px">
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Default col width",

  parameters: {
    layout: "fullscreen",
  },
};

export const DefaultColSpan = {
  render: () => html`
    <ic-layout-grid default-col-span="2">
      <div style="width: 100px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 80px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 80px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Default col span",

  parameters: {
    layout: "fullscreen",
  },
};

export const DefaultColSpanObject = {
  render: () => html`
    <ic-layout-grid default-col-span="2">
      <div style="width: 100px; height: 50px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 80px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 80px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
    <script>
      const grid = document.querySelector("ic-layout-grid");
      grid.defaultColSpan = { xs: 1, sm: 2, md: 2, lg: 4, xl: 4 };
    </script>
  `,

  name: "Default col span object",

  parameters: {
    layout: "fullscreen",
  },
};

export const DefaultRowSpan = {
  render: () => html`
    <ic-layout-grid default-row-span="2" rows="2">
      <div style="width: 50px; height: 100px; background-color: lightblue;">
        Grid item 1
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
    </ic-layout-grid>
  `,

  name: "Default row span",

  parameters: {
    layout: "fullscreen",
  },
};

export const GridItem = {
  render: () => html`
    <ic-layout-grid columns="4">
      <ic-layout-grid-item
        col-start="2"
        col-span="2"
        row-start="1"
        row-span="2"
      >
        <div style="width: 50px; height: 50px; background-color: lightblue;">
          Grid item 1
        </div>
      </ic-layout-grid-item>
      <ic-layout-grid-item hide-in-mobile-mode="true">
        <div style="width: 50px; height: 50px; background-color: lightblue;">
          Grid item 2
        </div>
      </ic-layout-grid-item>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 5
      </div>
    </ic-layout-grid>
  `,

  name: "Grid item",

  parameters: {
    layout: "fullscreen",
  },
};

const defaultArgs = {
  aligned: null,
  fullHeight: false,
  columns: 4,
  defaultColSpan: 1,
  defaultColWidth: "64px",
  defaultRowSpan: 1,
  gridSpacing: 2,
  rows: 1,
  type: "fixed",
  gridItemColSpan: 1,
  gridItemColStart: 1,
  gridItemRowSpan: 1,
  gridItemRowStart: 1,
  gridItemHideInMobileMode: false,
};

export const Playground = {
  render: (args) =>
    html`<ic-layout-grid
      style="border: 1px solid black;"
      columns=${args.columns}
      aligned=${args.aligned}
      full-height=${args.fullHeight}
      default-col-span=${args.defaultColSpan}
      default-col-width=${args.defaultColWidth}
      default-row-span=${args.defaultRowSpan}
      grid-spacing=${args.gridSpacing}
      rows=${args.rows}
      type=${args.type}
    >
      <ic-layout-grid-item
        col-span=${args.gridItemColSpan}
        col-start=${args.gridItemColStart}
        row-span=${args.gridItemRowSpan}
        row-start=${args.gridItemRowStart}
        hide-in-mobile-mode=${args.gridItemHideInMobileMode}
      >
        <div style="width: 50px; height: 50px; background-color: lightblue;">
          Grid item 1
        </div>
      </ic-layout-grid-item>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 2
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 3
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 4
      </div>
      <div style="width: 50px; height: 50px; background-color: lightblue;">
        Grid item 5
      </div>
    </ic-layout-grid>`,

  name: "Playground",
  args: defaultArgs,

  argTypes: {
    aligned: {
      options: [null, "left", "center", "full-width"],

      control: {
        type: "inline-radio",
      },
    },

    type: {
      options: ["fixed", "fluid"],

      control: {
        type: "inline-radio",
      },
    },
  },
};
