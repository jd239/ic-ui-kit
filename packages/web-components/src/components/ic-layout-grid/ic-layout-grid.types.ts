export type IcGridBreakpoints =
  | number
  | {
      xs?: number;
      sm?: number;
      md?: number;
      lg?: number;
      xl?: number;
    };

export type IcGridType = "fluid" | "fixed";
