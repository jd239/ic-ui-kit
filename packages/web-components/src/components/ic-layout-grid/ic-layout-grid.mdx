import { Meta, Stories, Markdown } from "@storybook/blocks";
import readme from "./readme.md";
import layoutGridItemReadme from "../ic-layout-grid-item/readme.md";
import * as IcLayoutGridStories from "./ic-layout-grid.stories";

<Meta of={IcLayoutGridStories} />

<Markdown>{readme + layoutGridItemReadme}</Markdown>

<Stories />
