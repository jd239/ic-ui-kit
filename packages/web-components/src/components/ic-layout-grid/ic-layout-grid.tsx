import { Component, Prop, Host, h, Element, Watch } from "@stencil/core";
import { IcAlignment } from "../../utils/types";
import { IcGridBreakpoints, IcGridType } from "./ic-layout-grid.types";

@Component({
  tag: "ic-layout-grid",
  styleUrl: "ic-layout-grid.css",
  shadow: true,
})
export class LayoutGrid {
  private COL_WIDTH_CSS_PROP = "--ic-layout-grid-col-width";

  @Element() el: HTMLIcLayoutGridElement;

  /**
   * The alignment of the grid.
   */
  @Prop() aligned?: IcAlignment = "left";

  /**
   * The number of columns in the grid.
   * This can be a number or an object specifying different values for each breakpoint.
   */
  @Prop() columns?: IcGridBreakpoints;
  @Watch("columns")
  watchColumns() {
    this.getColumns();
  }

  /**
   * The default column span for grid items.
   */
  @Prop() defaultColSpan?: IcGridBreakpoints = 1;
  @Watch("defaultColSpan")
  watchDefaultColSpan() {
    this.getColSpan();
  }

  /**
   * The default column width for grid items.
   * This is used when the grid type is set to "fixed".
   */
  @Prop() defaultColWidth?: string = "64px";
  @Watch("defaultColWidth")
  watchDefaultColWidth(newValue: string) {
    if (this.type === "fixed") {
      this.el.style.setProperty(this.COL_WIDTH_CSS_PROP, newValue);
    }
  }

  /**
   * The default row span for grid items.
   */
  @Prop() defaultRowSpan?: number = 1;
  @Watch("defaultRowSpan")
  watchDefaultRowSpan(newValue: number) {
    this.el.style.setProperty("--ic-layout-grid-row-span", `${newValue}`);
  }

  /**
   * If `true`, the standard vertical padding from the grid will be removed.
   */
  @Prop() fullHeight?: boolean = false;

  /**
   * The space between grid items, in multiples of the base spacing unit (`var(--ic-space-xs)`).
   */
  @Prop() gridSpacing?: number = 2;
  @Watch("gridSpacing")
  watchGridSpacing() {
    this.getGridSpacing();
  }

  /**
   * The number of rows in the grid.
   */
  @Prop() rows?: number = 1;
  @Watch("rows")
  watchRows(newValue: number) {
    this.el.style.setProperty("--ic-layout-grid-rows", `${newValue}`);
  }

  /**
   * The layout type of the grid.
   */
  @Prop() type: IcGridType = "fixed";
  @Watch("type")
  watchType(newValue: IcGridType) {
    this.getColumns();
    if (newValue === "fixed" && this.defaultColWidth) {
      this.el.style.setProperty(this.COL_WIDTH_CSS_PROP, this.defaultColWidth);
    } else {
      this.el.style.removeProperty(this.COL_WIDTH_CSS_PROP);
    }
  }

  componentWillLoad(): void {
    if (this.columns === undefined) {
      if (this.type === "fixed") {
        this.columns = 12;
      } else if (this.type === "fluid") {
        this.columns = {
          xs: 2,
          sm: 4,
          md: 8,
          lg: 12,
          xl: 12,
        };
      }
    }
    this.getColumns();
    this.getColSpan();
    this.getGridSpacing();
    if (this.type === "fixed" && this.defaultColWidth) {
      this.el.style.setProperty(this.COL_WIDTH_CSS_PROP, this.defaultColWidth);
    }
    if (this.rows) {
      this.el.style.setProperty("--ic-layout-grid-rows", `${this.rows}`);
      this.el.style.setProperty(
        "--ic-layout-grid-row-span",
        `${this.defaultRowSpan}`
      );
    }
  }

  private getColumns = () => {
    if (typeof this.columns === "number") {
      this.el.style.setProperty(
        "--ic-layout-grid-columns",
        this.columns.toString()
      );
    } else if (typeof this.columns === "object") {
      Object.entries(this.columns).forEach(([breakpoint, value]) => {
        if (value !== undefined) {
          this.el.style.setProperty(
            `--ic-layout-grid-columns-${breakpoint}`,
            value.toString()
          );
        }
      });
    }
  };

  private getColSpan = () => {
    if (this.defaultColSpan !== undefined) {
      if (typeof this.defaultColSpan === "number") {
        (Array.from(this.el.children) as HTMLElement[]).forEach((child) => {
          child.style.setProperty(
            "--ic-layout-grid-col-span",
            `${this.defaultColSpan}`
          );
        });
      } else if (typeof this.defaultColSpan === "object") {
        Object.entries(this.defaultColSpan).forEach(([breakpoint, value]) => {
          if (value !== undefined) {
            this.el.style.setProperty(
              `--ic-layout-grid-col-span-${breakpoint}`,
              value.toString()
            );
          }
        });
      }
    }
  };

  private getGridSpacing = () => {
    this.el.style.setProperty(
      "--ic-layout-grid-spacing",
      `calc(var(--ic-space-xs) * ${this.gridSpacing})`
    );
  };

  render() {
    const { aligned, fullHeight } = this;

    return (
      <Host
        class={{
          [`ic-layout-grid-aligned-${aligned}`]: true,
          ["ic-layout-grid-no-vertical-padding"]: !!fullHeight,
        }}
      >
        <slot></slot>
      </Host>
    );
  }
}
