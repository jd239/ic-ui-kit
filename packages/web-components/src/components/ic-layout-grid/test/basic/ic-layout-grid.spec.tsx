import { LayoutGrid } from "../../ic-layout-grid";
import { newSpecPage } from "@stencil/core/testing";

const COL_WIDTH_CSS_PROP = "--ic-layout-grid-col-width";

describe("ic-layout-grid", () => {
  it("should render with left-aligned by default", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid>
        <div>Test div 1</div>
        <div>Test div 2</div>
        <div>Test div 3</div>
        <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render full-width", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid aligned="full-width">
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render centered", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid aligned="center">
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should remove vertical padding when full-height", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid full-height>
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with columns set", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid columns="3">
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with rows set", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid rows="2">
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with defaultColSpan as a number", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid default-col-span="2">
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with defaultColSpan as an object", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid>
      <div>Test div 1</div>
      <div>Test div 2</div>
      <div>Test div 3</div>
      <div>Test div 4</div>
      </ic-layout-grid>`,
    });

    page.rootInstance.defaultColSpan = { xs: 2, sm: 3, md: 4, lg: 5, xl: 6 };

    expect(page.root).toMatchSnapshot();
  });

  it("should set CSS variables for columns at breakpoints", async () => {
    const columns = { xs: 2, sm: 4, md: 8, lg: 12, xl: 16 };
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid>
        <div>Test div 1</div>
        <div>Test div 2</div>
        <div>Test div 3</div>
        <div>Test div 4</div>
      </ic-layout-grid>`,
    });
    page.rootInstance.columns = columns;
    await page.rootInstance.getColumns();
    await page.waitForChanges();

    Object.entries(columns).forEach(([breakpoint, value]) => {
      expect(
        page.root?.style.getPropertyValue(
          `--ic-layout-grid-columns-${breakpoint}`
        )
      ).toBe(value.toString());
    });
  });

  it("should set CSS variables for defaultColSpan at breakpoints", async () => {
    const defaultColSpan = { xs: 2, sm: 3, md: 4, lg: 5, xl: 6 };
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid>
        <div>Test div 1</div>
        <div>Test div 2</div>
        <div>Test div 3</div>
        <div>Test div 4</div>
      </ic-layout-grid>`,
    });
    page.rootInstance.defaultColSpan = defaultColSpan;
    await page.rootInstance.getColSpan();
    await page.waitForChanges();

    Object.entries(defaultColSpan).forEach(([breakpoint, value]) => {
      expect(
        page.root?.style.getPropertyValue(
          `--ic-layout-grid-col-span-${breakpoint}`
        )
      ).toBe(value.toString());
    });
  });

  it("should update --ic-layout-grid-rows when rows changes", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid></ic-layout-grid>`,
    });
    page.rootInstance.rows = 5;
    await page.waitForChanges();
    expect(page.root!.style.getPropertyValue("--ic-layout-grid-rows")).toBe(
      "5"
    );
  });

  it("should update --ic-layout-grid-row-span when defaultRowSpan changes", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid></ic-layout-grid>`,
    });
    page.rootInstance.defaultRowSpan = 5;
    await page.waitForChanges();
    expect(page.root!.style.getPropertyValue("--ic-layout-grid-row-span")).toBe(
      "5"
    );
  });

  it("should update --ic-layout-grid-col-width when defaultColWidth changes", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid type="fixed"></ic-layout-grid>`,
    });
    page.rootInstance.defaultColWidth = 5;
    await page.waitForChanges();
    expect(page.root!.style.getPropertyValue(COL_WIDTH_CSS_PROP)).toBe("5");
  });

  it("should set and remove --ic-layout-grid-col-width when type changes", async () => {
    const page = await newSpecPage({
      components: [LayoutGrid],
      html: `<ic-layout-grid type="fixed" default-col-width="100px"></ic-layout-grid>`,
    });

    const grid = page.root;
    expect(grid!.style.getPropertyValue(COL_WIDTH_CSS_PROP)).toBe("100px");

    grid!.type = "fluid";
    await page.waitForChanges();
    expect(grid!.style.getPropertyValue(COL_WIDTH_CSS_PROP)).toBe("");

    grid!.type = "fixed";
    await page.waitForChanges();
    expect(grid!.style.getPropertyValue(COL_WIDTH_CSS_PROP)).toBe("100px");
  });
});
