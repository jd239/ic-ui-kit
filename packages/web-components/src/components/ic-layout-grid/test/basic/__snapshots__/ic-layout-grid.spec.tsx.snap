// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-layout-grid should remove vertical padding when full-height 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left ic-layout-grid-no-vertical-padding" full-height="" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render centered 1`] = `
<ic-layout-grid aligned="center" class="ic-layout-grid-aligned-center" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render full-width 1`] = `
<ic-layout-grid aligned="full-width" class="ic-layout-grid-aligned-full-width" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render with columns set 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left" columns="3" style="--ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render with defaultColSpan as a number 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left" default-col-span="2" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div>
    Test div 1
  </div>
  <div>
    Test div 2
  </div>
  <div>
    Test div 3
  </div>
  <div>
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render with defaultColSpan as an object 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1; --ic-layout-grid-col-span-xs: 2; --ic-layout-grid-col-span-sm: 3; --ic-layout-grid-col-span-md: 4; --ic-layout-grid-col-span-lg: 5; --ic-layout-grid-col-span-xl: 6;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render with left-aligned by default 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 1; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;

exports[`ic-layout-grid should render with rows set 1`] = `
<ic-layout-grid class="ic-layout-grid-aligned-left" rows="2" style="--ic-layout-grid-columns: 12; --ic-layout-grid-spacing: calc(var(--ic-space-xs) * 2); --ic-layout-grid-col-width: 64px; --ic-layout-grid-rows: 2; --ic-layout-grid-row-span: 1;">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 1
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 2
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 3
  </div>
  <div style="--ic-layout-grid-col-span: 1;">
    Test div 4
  </div>
</ic-layout-grid>
`;
